### DISCUSSION TITLE
Refactor or keep current? | 发布重构版或保持现状？

### DISCUSSION THREAD

**Discussion by crazywhalecc** [crazywhalecc/static-php-cli#29] [Announcements]
## 起因

因为作者长时间没有完善当前版本的项目了，因为本项目目前有很多局限性：

- 只能编译 Linux 环境的静态二进制文件。
- 添加新的扩展或依赖库不方便。
- 作者自己维护的资源镜像容易不更新。
- 不能导出和列明编译后所使用开源项目的 LICENSE，可能会造成违反开源协议的问题。

综上所述，作者在私有仓库维护了一个 PHP 编写的 static-php-cli 项目，用于多系统编译 PHP 和 micro 二进制，同时参考了 lwmbs、swoole-cli 和 phpmicro 项目的部分源码，就是作者正在编写的新版本。新版本具有可以拆装扩展和库、自动更新和打包资源等功能，同时几乎支持任意主流的 Linux、Windows 和 macOS。

但现有的 Bash 版也有一定的优势，就是简洁、易懂，无任何第三方库的使用，仅使用 *nix 工具集和 shell、Dockerfile 即可完成最小编译。

所以，现在征集一下各方的意见。如果同意作者发布重构版，可以数个大拇指，不同意的话就刷个火箭。

## Refactor reason

Because the author has not improved the current version of the project for a long time, the project has many limitations:

- Only static binaries for Linux environments can be compiled.
- It is inconvenient to add new extensions or dependent libraries.
- The libraries download mirror server maintained by the author is not updated frequently.
- The LICENSE of the open source project used after compilation cannot be exported and listed, which may cause a violation of the open source agreement.
- To sum up, the author maintains a static-php-cli project written in PHP in the private repo, which is used to compile PHP and micro binaries in multiple systems. At the same time, the author refers to some source codes of [lwmbs](/dixyes/lwmbs),[swoole-cli](/swoole/swoole-cli) and [phpmicro](/dixyes/phpmicro) projects, which is the new version being written by the author.

However, the existing Bash version also has certain advantages. It is concise and easy to understand, without the use of any third-party libraries, and only use *nix open source tools. By the way, shell, Dockerfile can complete the minimum compilation.

So now, let's gather the opinions of all parties. If you agree with the author to release the refactor version, you can give me a thumb. If you disagree, you can send me a rocket.

**Comment by ikilobyte** [comment]
支持之后直接发布新版本，我觉得当前版本可以不用维护了

**Comment by jingjingxyk** [comment]
创建一个新分支，并且设置为默认分支即可

```bash

git checkout  --orphan  static-php-cli-next

git rm -rf .


```

**Comment by TNT-ma** [comment]
我觉得这个玩法能不能实现一种创新方式
比如：先自己写php项目,写完后把自己项目打包为phar 然后在编译php二进制时,把自己的phar编译进去,最终形成一个真正的(linux) 二进制文件 ,而非单纯的一个cat合并搞一个 假二进制(假二进制无法使用upx等压缩壳软件)
当然我的方向是 往加密保户代码方向考虑

**Reply by crazywhalecc** [reply]
我的答案是，本项目永远不会考虑这个方向，首先 PHP 的 cPHP 是解释型语言，而且现在市面上还没有一个能兼容主流分支的 PHP 编译器。之前我在网上看到有一个项目是 BPC（Binary PHP Compiler？）可以把 PHP 编译为二进制，但是限制也是很多很多。

加密保护代码的方向和编译也不是一回事，编译过后也可以通过逆向工程等方式拿到代码，真正保护还是通过加壳、加密代码等手段进行。

所以本项目（static-php-cli）、相关项目（lwmbs、swoole-cli）都是提供一个对 php-src 源码的便捷编译工具，本项目和相关项目引用的 phpmicro 也仅仅是 PHP 的 sapi 接口封装，而不是 PHP 代码的编译工具。PHP 代码的编译器是完全不同的项目，因此不会考虑额外的情况。如果你对加密感兴趣，可以考虑使用现有的加密技术，如 Swoole Compiler、Source Guardian 等。

**Reply by TNT-ma** [reply]
明白 最近在看swoole

**Reply by dixyes** [reply]
加密自己用compiler啊，micro的目的不是保护代码
micro本身可以upx的 有个越南老哥试过: https://github.com/dongnguyenvt/phpmicro

**Reply by TNT-ma** [reply]
compiler  没有一个好用的

**Comment by crazywhalecc** [comment]
当前进度完成到了 macOS 平台的部分扩展，基础 CLI 框架已经编写完毕，但 Linux 系统下可能需要使用额外的 musl-libc，如果使用的系统没有 musl 时可能需要先编译安装 musl。

**Comment by crazywhalecc** [comment]
目前进度完成了 macOS 系统的静态 PHP 编译工作，同时适配了部分扩展。相关重构源码见 [refactor](https://github.com/crazywhalecc/static-php-cli/tree/refactor) 分支

**Comment by jingjingxyk** [comment]
耳目一新

**Comment by crazywhalecc** [comment]
目前提供了部分 Linux 的支持，但 Linux 环境有一定的要求，比如需要包管理安装 musl-libc、build-essential 等。

目前扩展支持情况：https://github.com/crazywhalecc/static-php-cli/blob/refactor/ext-support.md

**Reply by ikilobyte** [reply]
后续在linux上也会有这些要求吗

**Reply by crazywhalecc** [reply]
musl-libc 是 Linux 的基础 libc 库之一，这个是静态编译的硬性条件，暂无办法解决。build-essential 是一个编译工具包，包含 gcc 等命令，这些工具在 macOS 也同样需要，在 Windows 系统则也需要 nmake 等工具。理论上可以调用下载链接直接下载现成的，但最终还是要有编译环境，而且这时候用包管理装两个包远要比项目去兼容自动下载这些依赖来的容易。

Linux 编译环境可以通过 Docker 继续在 Alpine Linux 环境下编译，存在的问题会少很多，也可以在 Debian、Ubuntu 等环境安装编译工具和 musl 后编译。

**Reply by dixyes** [reply]
理论上来说 你是可以直接打包发布所有依赖的库和工具链的 类似国内各种平台的sdk 甚至有的直提供虚拟机镜像（点名批评esp）
做lwmbs时我发现要兼容各种环境太麻烦了 最后只支持了几个发行版

**Reply by crazywhalecc** [reply]
兼容环境确实是个大问题，但编译工具不能直接下载源码，只能下载对应系统的二进制，所以这里我觉得调用发行版的包管理来安装编译工具要更合适而且适配起来更容易一点？

**Reply by dixyes** [reply]
我就科普下国内有些厂商的操作

他们会把整个开发环境打包成虚拟机镜像，连带virtualbox一起打包

然后用户windows下载下来，双击打开就能用，这个肯定更方便使用，但是会很大
